### PR TITLE
vc4asm 0.1.7 (new formula)

### PR DESCRIPTION
**Pull request by fran6co**

### CODE REVIEW

File: Library/Formula/vc4asm.rb [hunk 1]
```diff
@@ -0,0 +1,16 @@
+class Vc4asm < Formula
+  homepage "http://maazl.de/project/vc4asm/doc/index.html"
+  url "https://github.com/maazl/vc4asm/archive/V0.1.7.tar.gz"
+  sha1 "20594439e9011dbd2175903754735cafbd01f029"
+
+  def install
+    rm_rf "bin"
```
Comment (DomT4) [inline Library/Formula/vc4asm.rb]: Why is there already a `bin` directory? Is there anything in it?

File: Library/Formula/vc4asm.rb [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+    mkdir "bin"
+    mkdir "obj"
```
Comment (DomT4) [inline Library/Formula/vc4asm.rb]: Why is this necessary?

File: Library/Formula/vc4asm.rb [hunk 3]
```diff
@@ -0,0 +1,16 @@
+class Vc4asm < Formula
+  homepage "http://maazl.de/project/vc4asm/doc/index.html"
+  url "https://github.com/maazl/vc4asm/archive/V0.1.7.tar.gz"
+  sha1 "20594439e9011dbd2175903754735cafbd01f029"
```
Comment (DomT4) [inline Library/Formula/vc4asm.rb]: `sha256` please.

File: Library/Formula/vc4asm.rb [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+    rm_rf "bin"
+    mkdir "bin"
+    mkdir "obj"
+    cd "src" do
+       system "make"
+    end
+    bin.install Dir["bin/*"]
+    share.install Dir["share/*"]
+  end
+end
```
Comment (DomT4) [inline Library/Formula/vc4asm.rb]: You'll need to add a `test do` block at the bottom here to test functionality of the tool after installation.

File: Library/Formula/vc4asm.rb (hunk 1 above)
Comment (fran6co) [inline Library/Formula/vc4asm.rb]: It has raspberry pi binaries, if I don't delete them it doesn't build it.

File: Library/Formula/vc4asm.rb (hunk 2 above)
Comment (fran6co) [inline Library/Formula/vc4asm.rb]: It doesn't create the obj folder on it's own. I'm working on PR for adding cmake support to the project as it should solve this most of this issues.

File: Library/Formula/vc4asm.rb (hunk 4 above)
Comment (fran6co) [inline Library/Formula/vc4asm.rb]: It doesn't have a nice testing suite, it has some assembly files that are supposed to fail but not sure if it makes sense here.

File: Library/Formula/vc4asm.rb (hunk 3 above)
Comment (fran6co) [inline Library/Formula/vc4asm.rb]: Done.

File: Library/Formula/vc4asm.rb (hunk 2 above)
Comment (DomT4) [inline Library/Formula/vc4asm.rb]: It'd be nice to get even a simple PR upstream to fix it not making a directory it needs to compile. Presuming they accepted it we could add it as a patch before the next release.

File: Library/Formula/vc4asm.rb (hunk 4 above)
Comment (DomT4) [inline Library/Formula/vc4asm.rb]: See `cmake.rb` for an example of an application formula with a good test and `tinyxml2.rb` for an example of a library formula with a good test - Try and replicate one of those. A simple test is better than no test.

Comment (fran6co) [inline Library/Formula/vc4asm.rb]: Done.

File: Library/Formula/vc4asm.rb [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+  homepage "http://maazl.de/project/vc4asm/doc/index.html"
+  url "https://github.com/maazl/vc4asm/archive/V0.1.7.tar.gz"
+  sha256 "db9c50c1cc035a183ce8305a82a01cad08a246d13c718c420a8762296b00e3de"
+
+  def install
+    rm_rf "bin"
+    mkdir "bin"
+    mkdir "obj"
+    cd "src" do
+       system "make"
```
Comment (bfontaine) [inline Library/Formula/vc4asm.rb]: Please use 2-spaces indentation here.

File: Library/Formula/vc4asm.rb [hunk 6]
```diff
@@ -0,0 +1,26 @@
+class Vc4asm < Formula
+  homepage "http://maazl.de/project/vc4asm/doc/index.html"
```
Comment (bfontaine) [inline Library/Formula/vc4asm.rb]: Could you please add a short description, using `desc "…"`? Thanks!

File: Library/Formula/vc4asm.rb (hunk 5 above)
Comment (fran6co) [inline Library/Formula/vc4asm.rb]: Done

File: Library/Formula/vc4asm.rb (hunk 6 above)
Comment (fran6co) [inline Library/Formula/vc4asm.rb]: Done

File: Library/Formula/vc4asm.rb [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+  desc "Macro assembler for Broadcom VideoCore IV aka Raspberry Pi GPU"
```
Comment (bfontaine) [inline Library/Formula/vc4asm.rb]: It should be above the homepage. Sorry for not mentioning this in my previous comment.

Comment (fran6co) [inline Library/Formula/vc4asm.rb]: No prob. Done

File: Library/Formula/vc4asm.rb (hunk 2 above)
Comment (fran6co) [inline Library/Formula/vc4asm.rb]: Made the PR upstream => https://github.com/maazl/vc4asm/pull/2

Using cmake solves this problems.

File: Library/Formula/vc4asm.rb [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+  patch do
+    url "https://github.com/maazl/vc4asm/pull/2.patch"
+    sha256 "9b7996563a6b15ae7d5578df6d08b43413dc1758e8e2002115ba414daed323e3"
+  end
+
+  def install
+    mkdir "build"
+    cd "build" do
+      system "cmake", "..", *std_cmake_args
+      system "make install"
```
Comment (bfontaine) [inline Library/Formula/vc4asm.rb]: Please use `system "make", "install"` here.

File: Library/Formula/vc4asm.rb [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+  end
+
+  test do
+    (testpath/"test.qasm").write <<-EOS.undent
+      mov -, sacq(9)
+      add r0, r4, ra1.unpack8b
+      add.unpack8a r0, r4, ra1
+      add r0, r4.8a, ra1
+    EOS
+    system "vc4asm", "-o test.hex", "-V", "#{share}/vc4.qinc", "test.qasm"
```
Comment (bfontaine) [inline Library/Formula/vc4asm.rb]: Use `"#{bin}/vc4asm"` instead of `"vc4asm"`.

File: Library/Formula/vc4asm.rb (hunk 8 above)
Comment (fran6co) [inline Library/Formula/vc4asm.rb]: Done.

File: Library/Formula/vc4asm.rb (hunk 9 above)
Comment (fran6co) [inline Library/Formula/vc4asm.rb]: Done.

**Comment by bfontaine** [comment]
The build [fails on Mountain Lion](http://bot.brew.sh/job/Homebrew%20Pull%20Requests/27621/version=mountain_lion/testReport/junit/brew-test-bot/mountain_lion/install_vc4asm/):

```
…snip…
Building CXX object src/CMakeFiles/vc4asm.dir/expr.cpp.o
cd /tmp/vc4asm20150610-15123-1oktng6/vc4asm-0.1.7/build/src && /usr/local/Library/ENV/4.3/clang++    -std=c++11   -o CMakeFiles/vc4dis.dir/utils.cpp.o -c /tmp/vc4asm20150610-15123-1oktng6/vc4asm-0.1.7/src/utils.cpp
cd /tmp/vc4asm20150610-15123-1oktng6/vc4asm-0.1.7/build/src && /usr/local/Library/ENV/4.3/clang++    -std=c++11   -o CMakeFiles/vc4asm.dir/expr.cpp.o -c /tmp/vc4asm20150610-15123-1oktng6/vc4asm-0.1.7/src/expr.cpp
In file included from /tmp/vc4asm20150610-15123-1oktng6/vc4asm-0.1.7/src/expr.cpp:8In file included from /tmp/vc4asm20150610-15123-1oktng6/vc4asm-0.1.7/src/expr.cpp:8:
/tmp/vc4asm20150610-15123-1oktng6/vc4asm-0.1.7/src/expr.h:13:10: fatal error: 'type_traits' file not found
:
/tmp/vc4asm20150610-15123-1oktng6/vc4asm-0.1.7/src/expr.h:13:10: fatal error: 'type_traits' file not found
#include <type_traits>
         ^
#include <type_traits>
         ^
1 error generated.
1 error generated.
make[2]: *** [src/CMakeFiles/vc4asm.dir/expr.cpp.o] Error 1
make[2]: *** [src/CMakeFiles/vc4dis.dir/expr.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: *** Waiting for unfinished jobs....
/tmp/vc4asm20150610-15123-1oktng6/vc4asm-0.1.7/src/utils.cpp/tmp/vc4asm20150610-15123-1oktng6/vc4asm-0.1.7/src/utils.cpp:39::3925:25: error: :no member named 'front' in 'std::basic_string<char>'
 error: {       if (rel.size() && rel.front() == '/')
                          ~~~ ^no member named 'front' in 'std::basic_string<char>'

{       if (rel.size() && rel.front() == '/')
                          ~~~ ^
/tmp/vc4asm20150610-15123-1oktng6/vc4asm-0.1.7/src/utils.cpp/tmp/vc4asm20150610-15123-1oktng6/vc4asm-0.1.7/src/utils.cpp::40:10: error: use of undeclared identifier 'move'
                return move(rel);
                       ^
40:10: error: use of undeclared identifier 'move'
                return move(rel);
                       ^
22 error errors generated.
```

**Comment by fran6co** [comment]
The problem is that vc4asm is a c++11 project and mountain lion doesn't support it as is. Not sure what is the correct procedure to fix this.

**Comment by DomT4** [comment]
> Not sure what is the correct procedure to fix this.

Add `needs :cxx11` to the main block, and `ENV.cxx11` to the top of `def install`.

**Comment by fran6co** [comment]
That fixed it!

File: Library/Formula/vc4asm.rb [hunk 10]
```diff
@@ -0,0 +1,32 @@
+class Vc4asm < Formula
+  desc "Macro assembler for Broadcom VideoCore IV aka Raspberry Pi GPU"
+  homepage "http://maazl.de/project/vc4asm/doc/index.html"
+  url "https://github.com/maazl/vc4asm/archive/V0.1.7.tar.gz"
+  sha256 "db9c50c1cc035a183ce8305a82a01cad08a246d13c718c420a8762296b00e3de"
+
+  depends_on "cmake" => :build
+
```
Comment (DomT4) [inline Library/Formula/vc4asm.rb]: Can you add a `needs :cxx11` on line 9 and push the patch down a couple of lines after, Thanks!

Comment (fran6co) [inline Library/Formula/vc4asm.rb]: Done.

File: Library/Formula/vc4asm.rb [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+  needs :cxx11
+
+
```
Comment (DomT4) [inline Library/Formula/vc4asm.rb]: Apologies, I meant the `patch do` line should migrate from line 9 to line 11 after the `needs` addition - Sorry I wasn't particularly clear on that.

Comment (fran6co) [inline Library/Formula/vc4asm.rb]: No prob.

**Comment by mikemcquaid** [comment]
Thanks for your contribution to Homebrew! Without people like you submitting PRs we couldn't run this project. You rock!
